planet/Objects.ocd/Environment.ocd/Cloud.ocd/Script.c

```diff
@@ -249,28 +280,136 @@ private func Precipitation()
 	return;
 }
 
-// Raindrop somewhere from the cloud.
-private func RainDrop()
+// Check if liquid is maybe in frozen form.
+private func RainMat()
 {
-	// Find Random Position.
-	var con = GetCon();
-	var wdt = GetDefWidth() * con / 500;
-	var hgt = GetDefHeight() * con / 700;
-	var x = RandomX(-wdt, wdt);
-	var y = RandomX(-hgt, hgt);
-	if (!GBackSky(x, y))
-		return false;
-	// Check if liquid is maybe in frozen form.
-	var mat;
 	if (rain_mat_freeze_temp != nil && GetTemperature() < rain_mat_freeze_temp)
-		mat = rain_mat_frozen;
+		return rain_mat_frozen;
 	else
-		mat = rain_mat;
-	// Create rain drop.	
-	CastPXS(mat, 1, 1, x, y);
+		return rain_mat;
+}
+
+local last_raindrop_color = -1;
+local particle_cache;
+
+// Raindrop somewhere from the cloud.
+private func RainDrop()
+{
+	var count = Max(rain_visual_strength, 1);
+	for (var i = 0; i < count; i++)
+	{
+		// Find Random Position.
+		var con = GetCon();
+		var wdt = GetDefWidth() * con / 500;
+		var hgt = GetDefHeight() * con / 700;
+		var x = RandomX(-wdt, wdt);
+		var y = RandomX(-hgt, hgt);
+		var xdir = RandomX(GetWind(0,0,1)-5, GetWind(0,0,1)+5)/5;
+		var ydir = 30;
+		if (!GBackSky(x, y))
+			continue;
+		var mat = RainMat();
+
+		if(mat == "Ice")
+		{
+			// Ice (-> hail) falls faster.
+			xdir *= 4;
+			ydir *= 4;
+		}
+
+		// Create rain particles.
+		var particle_name = "Raindrop";
+		var color = GetMaterialColor(mat);
+		if (color != last_raindrop_color)
+		{
+			particle_cache = {};
+			last_raindrop_color = color;
+		}
+
+		if(mat == "Lava" || mat == "DuroLava")
+			particle_name = "RaindropLava";
+
+		// Snow is special.
+		if(mat == "Snow")
+		{
+			particle_cache.snow = particle_cache.snow ?? Particles_Snow(color);
+			CreateParticle("RaindropSnow", x, y, xdir, 10, PV_Random(2000, 3000), particle_cache.snow, 0);
+			if (!i)
+			{
+				CastPXS(mat, 1, 0, x, y);
+			}
+			continue;
+		}		
+
+		particle_cache.rain = particle_cache.rain ?? Particles_Rain(color);
+		var particle = new particle_cache.rain {};
+		if(Random(2))
+			particle.Attach = ATTACH_Back;
+		CreateParticle(particle_name, x, y, xdir, ydir, PV_Random(200, 300), particle, 0);
+
+		// Splash.
+		var hit = SimFlight(x, y, xdir, ydir, 25 /* Liquid */, nil, nil, 3);
+		var x_final = hit[0], y_final = hit[1], time_passed = hit[4];
+		if (time_passed > 0)
+		{
+			// Don't always create a splash.
+			if (!Random(5) || !i)
+				ScheduleCall(this, "DropHit", time_passed, 0, mat, color, AbsX(x_final), AbsY(y_final), !i);
+		}
+	}
 	return true;
 }
 
+private func DropHit(string material_name, int color, int x_orig, int y_orig, bool create_material)
+{
+	// Adjust position so that it's in the air.
+	var x = x_orig, y = y_orig;
+	while (GBackSemiSolid(x, y - 1)) y--;
+
+	if (create_material)
+	{
+		InsertMaterial(Material(material_name), x, y);
+	}
+
+	// Some material combinations cast smoke
+	// TODO: Figure out some generic way to do this?
```

Comment:
Well, which material combinations are supposed to cast smoke? Right now, there's nothing that can be genericized because it looks like an arbitrary list.